### DISCUSSION TITLE
run scripts when phase1 succeeds

### DISCUSSION THREAD

**Discussion by shiningw** [strongswan/strongswan#1438] [Ideas]
Hi,

Can I run some scripts when phase1 ends and phase2 is about to start using udp 4500? If not, where can I do the hacking inside strongswan? What I want to achieve is to open port 4500 when phase1 succeeds. Thank you!

**Comment by strongX509** [comment]
`phase1` and `phase2` are actually concepts going back to `IKEv1` but with `IKEv2` you could say `phase1` establishes an `IKE_SA` whereas `phase2` establishes a `CHILD_SA`.

Establishment of an `IKE_SA` consists of the exchange of the `IKE_SA_INIT` and `IKE_AUTH` request/response message pairs where the floating to UDP port 4500 happens when the `IKE_SA_INIT` exchange detects a NAT situation or if `MOBIKE` is enabled. Thus UDP port 4500 is already needed to complete `phase1`.

The IKEv2 RFC 7296 even states in the[ NAT-Traversal section 2.23](https://datatracker.ietf.org/doc/html/rfc7296#section-2.23):

> An initiator can use port 4500 for both IKE and ESP, regardless of
>  whether or not there is a NAT, even at the beginning of IKE. 

Further I don't understand how selectively blocking/unblocking UDP port 4500 would increase the security of your system since UDP port 500 has to be open all the time anyway and with your proposed approach any attacker could open UDP port 4500 on the peer by just sending an unauthenticated `IKE_SA_INIT` request.

**Comment by shiningw** [comment]
@strongX509 Thanks for your reply. I want to do this because I suspect the government firewall is blocking UDP 4500 when it detects this port is open. Instead of allowing the firewall to probe my server easily, I want to make it a little more difficult by keeping this port open to only those who can at least complete the phase 1 authentication. 
    My connection usually get stuck at the end of phase 1 after a day or two of usage. But I can connect normally again after switching nat-t port to a non-standard udp port . that's why I suspect that keeping the port closed may help avoid the attention of the national firewall. 
      I have studied strongswan's source code for some hours but still not sure where to put the modification. Please point me to the right directon where I can modify the code to get this done. I have basic knowledge of C language. Thank you!

**Reply by strongX509** [reply]
You could implement a [listener](https://github.com/strongswan/strongswan/blob/master/src/libcharon/bus/listeners/listener.h#L46) in a proprietary libcharon plugin which subscribes to an IKE_SA state change (up|down) on the internal bus but as I have written above you already need port 4500 in order for the IKE_SA (phase1) to complete.

**Reply by shiningw** [reply]
> You could implement a [listener](https://github.com/strongswan/strongswan/blob/master/src/libcharon/bus/listeners/listener.h#L46) in a proprietary libcharon plugin which subscribes to an IKE_SA state change (up|down) on the internal bus but as I have written above you already need port 4500 in order for the IKE_SA (phase1) to complete.

Ok. What about making sure that the packet is coming from an IKE client instead of some port probing programs? just execute a firewall rule towards the end of phase 1. the only parameter I will need is the source IP. Thank you!

**Reply by shiningw** [reply]
> You could implement a [listener](https://github.com/strongswan/strongswan/blob/master/src/libcharon/bus/listeners/listener.h#L46) in a proprietary libcharon plugin which subscribes to an IKE_SA state change (up|down) on the internal bus but as I have written above you already need port 4500 in order for the IKE_SA (phase1) to complete.

I'm looking at here https://github.com/strongswan/strongswan/blob/4dd3d0e57bd44f792324a496692ad8caffe0720e/src/libcharon/sa/ikev1/phase1.c#L566 
can you tell me how to get the client ip here? Thank you!

**Reply by strongX509** [reply]
`other->get_address(other)` returns the IP address of the peer as a  binary `chunk_t` with a length of 4 bytes

**Reply by shiningw** [reply]
> `other->get_address(other)` returns the IP address of the peer as a binary `chunk_t` with a length of 4 bytes

I figured it out! many thanks for your help!
